scripts/source-maintenance.sh

```diff
@@ -66,7 +86,11 @@ for FILENAME in `git ls-files`; do
 	#
 	# Code Style formatting maintenance
 	#
-        if test "${ASVER}"; then
+	for SCRIPT in `git ls-files scripts/maintenance/`; do
+		${SCRIPT} "${FILENAME}" >"${FILENAME}.new" &&
+			updateifchanged "${FILENAME}" "${FILENAME}.new" "by ${SCRIPT}"
+	done
```

Comment:
The approach in this PR does not protect unstaged changes from being wiped out by source-maintenance.sh. We should not accept such risks IMO, especially when we are adding generic "plugin" support. I think we should adopt scripts/spell-check.sh approach: Refuse to adjust any files in a repository with unstaged changes. It is trivial to implement. If that simple approach has problems in source-maintenance.sh context, let's discuss.

Reply:
This request is effectively a "new feature request" for the maintenance script. Outside the scope of this PR.

The source-maintenance.sh as a whole is intentionally capable of reformatting any file in the repository, including staged ones. Specifically so that files can be staged, have the formatting enforcement run to check them before merge.  "Blowing away" a staged file shows up in the resulting diff/patch as clearly as the same action on a committed file - that case has been tested for long ago.

Reply:
I will reorder your paragraphs because there might simply be a misunderstanding here. The added emphasis is mine. I hope that the second set of paragraphs (i.e. "... including staged ones") will clarify things.

> The source-maintenance.sh as a whole is intentionally capable of reformatting _any_ file in the repository

Agreed.

I am not sure whether the following minor aspect is a part of the bigger misunderstanding, but just in case: If there are valid reasons for doing so, I will happily agree to grandfather existing source-maintenance.sh modification avenues, applying the suggested no-unstaged-changes restriction just to _plugins_. The difference between "the restriction applies to all source-maintenance.sh changes" and "the restriction applies to changes by plugins" is not critically important to me for this PR. It would be simpler to apply the restriction globally, but if there are valid reasons for a narrower application, they can be accommodated AFAICT.

> ..., including _staged_ ones. Specifically so that files can be staged, have the formatting enforcement run to check them before merge. "Blowing away" a staged file shows up in the resulting diff/patch as clearly as the same action on a committed file - that case has been tested for long ago.

Exactly! We seem to be in an agreement here! Perhaps you thought that I am asking you to prohibit applying changes to staged files? I am _not_ asking for that, and asking for that would be _wrong_. There is nothing wrong with modifying a staged file. It is a safe operation.

To reiterate this change request: I think we should prohibit script-driven modification of files with _unstaged_ changes, and I am suggesting accomplishing that by applying the restriction to the repository as a whole (as already implemented in the official scripts/spell-check.sh):

```shell
if ! git diff --quiet; then
    echo "There are unstaged changes. This script may modify sources."
    echo "Stage changes to avoid permanent losses when things go bad."
    exit 1
fi
```

> This request is effectively a "new feature request" for the maintenance script. Outside the scope of this PR.

I disagree on both counts: IMO, the requested core functionality is _required_ to properly support "running of arbitrary code" -- a stated PR goal. The new-out-of-scope-feature argument could be valid if we were just polishing the existing script to do better what it already does. This PR adds a new powerful mechanism. I can support the addition of that mechanism if it has certain safety guarantees. Fortunately, adding those guarantees is very simple (see above for a sketch).

BTW, I was just told that avoiding script-changing unstaged files is a popular/common approach in git days. We are not inventing anything new or doing anything very controversial here! It is a natural solution for mitigating a real risk.

Reply:
I am doing this, but only because it is not worth arguing over. I stand by it being feature creep on the current PR scope.